### PR TITLE
Automated allocation of components to target environments

### PR DESCRIPTION
**Pull request by adrianwersching**
Automated method to select best suited cloud provider (target location) for components of existing topologies.
Providers can be selected based on three criteria:
- Min Hosts: reduce amount of host components in the topology and prefer providers with matching PaaS
- Fulfill Policies: select providers by needed non-functional requirements (modeled as policies)
- Min External Connections: reduce connectsTo relations between different providers

Each criteria creates possible topology results and these results can be reduced by filtering them based on other criteria.

39956988-76ae9d08-55eb-11e8-870a-fb0752a0b1af
40133685-c5da5edc-5940-11e8-950c-4c4209d3ddd4

- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request. Especially, we require **a single commit**
- [x] Ensure that the commit message is [a good commit message](https://chris.beams.io/posts/git-commit/)
- [x] Ensure to use auto format in **all** files
- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [x] Screenshots added (for UI changes)

### CODE REVIEW

**Comment by genie-winery** [comment]
Can one of the admins verify this patch?

**Comment by adrianwersching** [comment]
Note: ServiceTemplateResourceTest.getVersionDifferences() is failing at the moment

**Comment by koppor** [comment]
add to whitelist

File: org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/policySelection/policySelection.component.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    templateUrl: 'policySelection.component.html'
+})
+
+export class PolicySelectionComponent implements OnInit {
+
+    @ViewChild('property') propertySelect: SelectComponent;
+
+    // policy, property, operator selection
+    policiesForSelect: Array<any> = [];
+    propertiesForSelect: Array<any> = [];
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/policySelection/policySelection.component.ts]: `{ id: key, text: key }` is used here: you could reuse the `SelectItem` interface here

File: org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/targetAllocation.component.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    templateUrl: 'targetAllocation.component.html'
+})
+
+export class TargetAllocationComponent implements OnInit {
+
+    @ViewChild(ModalDirective) targetAllocationModal: ModalDirective;
+    @ViewChild(PolicySelectionComponent) policySelection: PolicySelectionComponent;
+
+    // criteria selection spinner
+    criteria = [
```
Comment (lharzenetter) [inline org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/targetAllocation.component.ts]: Could also be of type `SelectItem`

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java [hunk 3]
```diff
@@ -744,8 +764,21 @@ public TTopologyTemplate injectNodeTemplates(TTopologyTemplate topologyTemplate,
             }
         }
 
-        Map<TNodeTemplate, Set<TNodeTemplate>> transitiveAndDirectSuccessors = computeTransitiveClosure(topologyTemplate);
+        switch (removal) {
+            case REMOVE_NOTHING:
+                return topologyTemplate;
+            case REMOVE_REPLACED:
+                for (TNodeTemplate deleteOriginNode : replacedNodeTemplatesToDelete) {
+                    topologyTemplate.getNodeTemplateOrRelationshipTemplate().removeAll(ModelUtilities.getIncomingRelationshipTemplates(topologyTemplate, deleteOriginNode));
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java]: Please reformat code fragments like these into multiple lines which enables a more readable code :)

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java [hunk 4]
```diff
@@ -744,8 +764,21 @@ public TTopologyTemplate injectNodeTemplates(TTopologyTemplate topologyTemplate,
...
+                    topologyTemplate.getNodeTemplateOrRelationshipTemplate().removeAll(ModelUtilities.getOutgoingRelationshipTemplates(topologyTemplate, deleteOriginNode));
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java]: Please reformat code fragments like these into multiple lines which enables a more readable code :)

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java [hunk 5]
```diff
@@ -1314,4 +1345,10 @@ private void addMatchingRelationshipTemplateToTopologyTemplate(TTopologyTemplate
         matchingRelationshipTemplate.setTargetElement(targetElement);
         topologyTemplate.getNodeTemplateOrRelationshipTemplate().add(matchingRelationshipTemplate);
     }
+
+    public enum InjectRemoval {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java]: Please move this Enum to a separate file - this file is already large enough...

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/targetallocation/criteria/minhosts/MinHosts.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+     * and selecting first match.
+     */
+    @Override
+    protected List<List<PermutationHelper>> getPossibleMatches(TopologyWrapper topology) throws AllocationException {
+        Map<TNodeTemplate, List<TTopologyTemplate>> possibleReplacements = new HashMap<>();
+        List<TNodeTemplate> done = new ArrayList<>();
+
+        for (TNodeTemplate topLevelHost : topology.getTopLevelHosts()) {
+            TNodeTemplate next = topLevelHost;
+            String targetLabel = ModelUtilities.getTargetLabel(topology.getHostedOnPredecessors(topLevelHost).get(0)).get();
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/targetallocation/criteria/minhosts/MinHosts.java]: What if `... .get(0)).get()` is `null`?

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/targetallocation/criteria/minhosts/TargetLabelAssignment.java [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    private List<TNodeTemplate> addTargetLabelsToTransitiveTopLevelNTPredecessors(TNodeTemplate nodeTemplate,
+                                                                                  Set<String> targetLabels) {
+        List<TNodeTemplate> topLevelPredecessors = original.getTransitiveTopLevelPredecessors(nodeTemplate);
+        for (TNodeTemplate topLevelPredecessor : topLevelPredecessors) {
+            possibleTargetLabels.get(topLevelPredecessor).addAll(targetLabels);
+        }
+        return topLevelPredecessors;
+    }
+
+    private enum PredecessorsTargetLabelState {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/targetallocation/criteria/minhosts/TargetLabelAssignment.java]: Please move this enum to the top of the file

File: org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/KargerMinCutVariationTest.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(KargerMinCutVariationTest.class);
+
+    @BeforeEach
+    public void setUp() throws GitAPIException {
+        setRevisionTo("bb9ff7e08f7b72d30a2fae326740ef8051701671");
+    }
+
+    @Test
+    public void test1() {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/KargerMinCutVariationTest.java]: Please rename the test to cleraly descripe what it does. Otherwiese it is more difficult to understand it.

File: org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/KargerMinCutVariationTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        ConnectsToGraph.Node pastureNode = result.getNodes().stream()
+            .filter(node -> node.getTargetLabel().equalsIgnoreCase("PastureProvider")).collect(Collectors.toList()).get(0);
+        ConnectsToGraph.Node fieldNode = result.getNodes().stream()
+            .filter(node -> node.getTargetLabel().equalsIgnoreCase("FieldProvider")).collect(Collectors.toList()).get(0);
+        assertEquals(pastureNode.getNodeTemplateIds(), pastureProvider);
+        assertEquals(fieldNode.getNodeTemplateIds(), fieldProvider);
+    }
+
+    @Test
+    public void test2() {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/KargerMinCutVariationTest.java]: Please rename the test to cleraly descripe what it does. Otherwiese it is more difficult to understand it.

File: org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/MinExternalConnectionsTest.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+    private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(MinExternalConnectionsTest.class);
+
+    @BeforeEach
+    public void setUp() throws GitAPIException {
+        setRevisionTo("bb9ff7e08f7b72d30a2fae326740ef8051701671");
+    }
+
+    @Test
+    public void allocate1() throws Exception {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/MinExternalConnectionsTest.java]: Please rename the test to cleraly descripe what it does. Otherwiese it is more difficult to understand it.

File: org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/MinExternalConnectionsTest.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        if (topologies.size() != 1) {
+            LOGGER.debug("Randomized algorithm didn't find correct results");
+            return;
+        }
+
+        assertEquals(6, topologies.get(0).getNodeTemplates().size());
+    }
+
+    @Test
+    public void allocate2() throws Exception {
```
Comment (lharzenetter) [inline org.eclipse.winery.repository/src/test/java/org/eclipse/winery/repository/targetallocation/criteria/minexternalconnections/MinExternalConnectionsTest.java]: Please rename the test to cleraly descripe what it does. Otherwiese it is more difficult to understand it.

File: org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/request.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+export class AllocationRequest {
+    assignOnly = false;
+    outputCap = 10;
+    selectedCriteria: Array<CriteriaRequest> = [];
+}
+
+export class CriteriaRequest {
+    criteria: string;
+    criteriaParams: any;
```
[the remark this answers is not in the document]
Comment (adrianwersching) [reply org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/request.ts]: The criteria are supposed to be plugin based and can have different parameters. And with generics I would have a heterogenous array.

File: org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/policySelection/policySelection.component.ts [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+                policy: this.policy,
+                property: this.policyProperty,
+                operator: this.operator
+            });
+        } else {
+            this.notify.error('Please enter all parameters');
+        }
+    }
+
+    onDelete(row: any) {
```
[the remark this answers is not in the document]
Comment (adrianwersching) [reply org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/policySelection/policySelection.component.ts]: Unfortunately I don't

[the remark this answers is not in the document]
Comment (koppor) [reply org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/policySelection/policySelection.component.ts]: Could you just add a comment why not? - No need to create a full [ADR](http://adr.github.io/madr/); a simple comment is enough.

File: org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/request.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (koppor) [reply org.eclipse.winery.repository.ui/src/app/wineryTargetAllocation/request.ts]: Could you please add a comment here? - Reason: The code should be self-documented and one should not need to have the knowledge of pull request comments to understand it. 😇

**Comment by koppor** [comment]
Sorry for working on BackendUtils.java again - could you please merge upstream/master again?

Could you please make on commit so that we can move on if nothing is left. I think, fixing the conflicts and add two code comments are the only things left?

**Comment by adrianwersching** [comment]
I added the comments, resolved the conflict and squashed the commits.

**Comment by adrianwersching** [comment]
ServiceTemplateResource.createNewVersion() is currently failing

**Comment by adrianwersching** [comment]
I checked out eclipse/master and its failing there too, so it should not be connected to my changes.

**Comment by adrianwersching** [comment]
I'm on holiday at the moment, I will resolve the conflicts when I return.

**Comment by adrianwersching** [comment]
I resolved the conflicts but there seems to be a problem with the winery-verification yaml.converter build that I can't reproduce locally. I haven't changed anything in the converter, is this a known issue?

**Comment by lharzenetter** [comment]
Did you merge the current master?

**Comment by lharzenetter** [comment]
Yes, this is okay. Travis completes successfully which is OK.

**Comment by saatkamp** [comment]
Filed fb00214.. as [CQ 17254](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=17254)

**Comment by lharzenetter** [comment]
Merging based on parallel IP.

**Pull request merged**